registry/registry.go

```diff
@@ -0,0 +1,178 @@
+package registry
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"unicode"
+
+	// thanks, go-digest...
+	_ "crypto/sha256"
+	_ "crypto/sha512"
+
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/reference/docker"
+	"github.com/containerd/containerd/remotes"
+	dockerremote "github.com/containerd/containerd/remotes/docker"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type ResolvedImage struct {
+	Ref  string
+	Desc ocispec.Descriptor
+
+	resolver remotes.Resolver
+	fetcher  remotes.Fetcher
+}
+
+func (img ResolvedImage) FetchJSON(ctx context.Context, v interface{}) error {
+	// prevent go-digest panics later
+	if err := img.Desc.Digest.Validate(); err != nil {
+		return err
+	}
+
+	r, err := img.fetcher.Fetch(ctx, img.Desc)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+
+	// make sure we can't possibly read (much) more than we're supposed to
+	limited := &io.LimitedReader{
+		R: r,
+		N: img.Desc.Size + 1, // +1 to allow us to detect if we read too much (see verification below)
+	}
+
+	// copy all read data into the digest verifier so we can validate afterwards
+	verifier := img.Desc.Digest.Verifier()
+	tee := io.TeeReader(limited, verifier)
+
+	// decode directly! (mostly avoids double memory hit for big objects)
+	// (TODO protect against malicious objects somehow?)
+	if err := json.NewDecoder(tee).Decode(v); err != nil {
+		return err
+	}
+
+	// read anything leftover ...
+	bs, err := io.ReadAll(tee)
+	if err != nil {
+		return err
+	}
+	// ... and make sure it was just whitespace, if anything
+	for _, b := range bs {
+		if !unicode.IsSpace(rune(b)) {
+			return fmt.Errorf("unexpected non-whitespace at the end of %q: %+v\n", img.Desc.Digest.String(), rune(b))
+		}
+	}
+
+	// after reading *everything*, we should have exactly one byte left in our LimitedReader (anything else is an error)
+	if limited.N < 1 {
+		return fmt.Errorf("size of %q is bigger than it should be (%d)", img.Desc.Digest.String(), img.Desc.Size)
+	} else if limited.N > 1 {
+		return fmt.Errorf("size of %q is %d bytes smaller than it should be (%d)", img.Desc.Digest.String(), limited.N-1, img.Desc.Size)
+	}
+
+	// and finally, let's verify our checksum
+	if !verifier.Verified() {
+		return fmt.Errorf("digest of %q not correct", img.Desc.Digest.String())
+	}
+
+	return nil
+}
+
+func (img ResolvedImage) Manifests(ctx context.Context) ([]ocispec.Descriptor, error) {
+	if img.IsImageManifest() {
+		return []ocispec.Descriptor{img.Desc}, nil
+	}
+
+	if !img.IsImageIndex() {
+		return nil, fmt.Errorf("unknown media type: %q", img.Desc.MediaType)
+	}
+
+	// (perhaps use a containerd content store??)
+	var index ocispec.Index
+	if err := img.FetchJSON(ctx, &index); err != nil {
+		return nil, err
+	}
+	return index.Manifests, nil
+}
+
+func (img ResolvedImage) Manifest(ctx context.Context, desc ocispec.Descriptor) (*ocispec.Manifest, error) {
+	img.Desc = desc // TODO this is a hack 🙈
```

Comment:
If I `s/ResolvedImage/ResolvedObject/g; s/img/obj/g` then this isn't a hack as much as Working-As-Designed :thinking: